File: ios/App/App/NativeFeatureFlags.swift
import Foundation

/// Native-side kill switches. Defaults OFF — hybrid WKWebView remains default.
enum NativeFeatureFlags {
    private static let suite = UserDefaults.standard
    private static let keyHost = "FF_NATIVE_HOST"
    private static let keyTrack = "FF_NATIVE_TRACK"
    private static let keyPct = "FF_NATIVE_TRACK_PCT"
    private static let keyEmails = "FF_NATIVE_TRACK_INTERNAL_EMAILS"
    private static let keyApiBase = "api_base_url"
    private static let keyApiVersion = "api_version"

    static var isNativeHostEnabled: Bool { suite.bool(forKey: keyHost) }
    static var isNativeTrackEnabled: Bool { suite.bool(forKey: keyTrack) }
    static var trackPercent: Int { min(100, max(0, suite.integer(forKey: keyPct))) }

    static var internalEmails: Set<String> {
        let raw = suite.string(forKey: keyEmails) ?? ""
        return Set(
            raw.split(separator: ",")
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() }
                .filter { !$0.isEmpty }
        )
    }

    static var apiBaseUrl: String {
        var value = suite.string(forKey: keyApiBase) ?? "https://manaintibojanam-backend.onrender.com"
        while value.hasSuffix("/") { value.removeLast() }
        return value
    }

    static var apiVersion: String {
        let value = suite.string(forKey: keyApiVersion) ?? "1.0"
        return value.isEmpty ? "1.0" : value
    }

    static func configure(
        nativeHost: Bool,
        nativeTrack: Bool,
        percent: Int,
        internalEmails: String,
        apiBaseUrl: String,
        apiVersion: String
    ) {
        suite.set(nativeHost, forKey: keyHost)
        suite.set(nativeTrack, forKey: keyTrack)
        suite.set(min(100, max(0, percent)), forKey: keyPct)
        suite.set(internalEmails, forKey: keyEmails)
        var base = apiBaseUrl
        while base.hasSuffix("/") { base.removeLast() }
        suite.set(base, forKey: keyApiBase)
        suite.set(apiVersion.isEmpty ? "1.0" : apiVersion, forKey: keyApiVersion)
    }
}

File: android/app/src/main/java/com/bhojanos/orderbhojan/nativehost/NativeFeatureFlags.kt
package com.bhojanos.orderbhojan.nativehost

import android.content.Context

/**
 * Native-side kill switches. Defaults OFF — hybrid WebView remains default.
 * Synced from JS via OrderBhojanNativeTrack.configure.
 */
object NativeFeatureFlags {
    private const val PREFS = "ob_native_flags"
    private const val KEY_HOST = "FF_NATIVE_HOST"
    private const val KEY_TRACK = "FF_NATIVE_TRACK"
    private const val KEY_PCT = "FF_NATIVE_TRACK_PCT"
    private const val KEY_EMAILS = "FF_NATIVE_TRACK_INTERNAL_EMAILS"
    private const val KEY_API_BASE = "api_base_url"
    private const val KEY_API_VERSION = "api_version"

    fun isNativeHostEnabled(context: Context): Boolean =
        prefs(context).getBoolean(KEY_HOST, false)

    fun isNativeTrackEnabled(context: Context): Boolean =
        prefs(context).getBoolean(KEY_TRACK, false)

    fun trackPercent(context: Context): Int =
        prefs(context).getInt(KEY_PCT, 0).coerceIn(0, 100)

    fun internalEmails(context: Context): Set<String> =
        prefs(context)
            .getString(KEY_EMAILS, "")
            .orEmpty()
            .split(',')
            .map { it.trim().lowercase() }
            .filter { it.isNotEmpty() }
            .toSet()

    fun apiBaseUrl(context: Context): String =
        prefs(context).getString(KEY_API_BASE, "https://manaintibojanam-backend.onrender.com")
            .orEmpty()
            .trimEnd('/')

    fun apiVersion(context: Context): String =
        prefs(context).getString(KEY_API_VERSION, "1.0") ?: "1.0"

    fun configure(
        context: Context,
        nativeHost: Boolean,
        nativeTrack: Boolean,
        percent: Int,
        internalEmails: String,
        apiBaseUrl: String,
        apiVersion: String,
    ) {
        prefs(context).edit()
            .putBoolean(KEY_HOST, nativeHost)
            .putBoolean(KEY_TRACK, nativeTrack)
            .putInt(KEY_PCT, percent.coerceIn(0, 100))
            .putString(KEY_EMAILS, internalEmails)
            .putString(KEY_API_BASE, apiBaseUrl.trimEnd('/'))
            .putString(KEY_API_VERSION, apiVersion.ifBlank { "1.0" })
            .apply()
    }

    private fun prefs(context: Context) =
        context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
}
